hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/OneReplicaPipelineSafeModeRule.java

```diff
@@ -171,6 +184,25 @@ public synchronized void refresh(boolean forceRefresh) {
     }
   }
 
+  private void updateReportedPipelineSet() {
+    List<Pipeline> openRatisPipelines =
+        pipelineManager.getPipelines(RatisReplicationConfig.getInstance(ReplicationFactor.THREE),
+            Pipeline.PipelineState.OPEN);
+
+    for (Pipeline pipeline : openRatisPipelines) {
+      PipelineID pipelineID = pipeline.getId();
+      boolean atleastOnePipelineReported = !pipeline.getNodeSet().isEmpty();
+      boolean notAlreadyReported = !reportedPipelineIDSet.contains(pipelineID);
+      boolean wasExistingPipeline = oldPipelineIDSet.contains(pipelineID);
+
+      if (atleastOnePipelineReported && notAlreadyReported && wasExistingPipeline) {
+        getSafeModeMetrics().incCurrentHealthyPipelinesWithAtleastOneReplicaReportedCount();
+        currentReportedPipelineCount++;
+        reportedPipelineIDSet.add(pipelineID);
```

Comment:
- Please inline the variables to avoid eagerly checking all 3 conditions.
- nit: we can skip `!reportedPipelineIDSet.contains(pipelineID)` by adding `reportedPipelineIDSet.add(pipelineID)` as last condition

```suggestion
      if (!pipeline.getNodeSet().isEmpty()
          && oldPipelineIDSet.contains(pipelineID)
          && reportedPipelineIDSet.add(pipelineID)) {
       getSafeModeMetrics().incCurrentHealthyPipelinesWithAtleastOneReplicaReportedCount();
        currentReportedPipelineCount++;
```